### PR TITLE
Laboratory work #4, Diana Gribanova - 22FPL1

### PR DESCRIPTION
**Pull request by Dzianna7**

### CODE REVIEW

**Review by marina-kaz — COMMENTED** [review]
Nothing to review just yet

File: lab_4_fill_words_by_ngrams/main.py [hunk 1]
```diff
@@ -28,6 +28,16 @@ def _tokenize(self, text: str) -> tuple[str, ...]:  # type: ignore
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if not isinstance(text, str) or len(text) == 0:
```
Comment (marina-kaz) [inline lab_4_fill_words_by_ngrams/main.py:31]: just use `not text`

**Review by marina-kaz — COMMENTED** [review]
It is a start, keep going

**Review by marina-kaz — COMMENTED** [review]
No new changes to review

File: lab_4_fill_words_by_ngrams/main.py [hunk 2]
```diff
@@ -28,6 +29,18 @@ def _tokenize(self, text: str) -> tuple[str, ...]:  # type: ignore
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if not isinstance(text, str) or not text:
+            raise ValueError
+        text = text.lower().split()
+        tokenized_text = []
+        for word in text:
+            if word[-1] in ('.', '!', '?'):
+                tokenized_text.extend([word[:-1], self._end_of_word_token])
```
Comment (marina-kaz) [inline lab_4_fill_words_by_ngrams/main.py:38]: no cleaning for these chunks?

File: lab_4_fill_words_by_ngrams/main.py [hunk 3]
```diff
@@ -56,6 +74,17 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> str:  #
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if not isinstance(decoded_corpus, tuple) or not decoded_corpus:
+            raise ValueError
+        decoded_corpus = ' '.join(decoded_corpus).replace(f' {self._end_of_word_token}', '.').capitalize()
+        capitalized_text = decoded_corpus.split('. ')
+        for i in range(1, len(capitalized_text)):
+            capitalized_text[i] = capitalized_text[i][0].upper() + capitalized_text[i][1:]
```
Comment (marina-kaz) [inline lab_4_fill_words_by_ngrams/main.py:82]: why use upper on a part of element when you could use capitalize on the whole element?

File: lab_4_fill_words_by_ngrams/main.py [hunk 4]
```diff
@@ -56,6 +74,17 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> str:  #
...
+        final_text = '. '.join(capitalized_text)
+        if final_text.endswith("."):
+            return final_text
+        else:
+            return final_text + '.'
```
Comment (marina-kaz) [inline lab_4_fill_words_by_ngrams/main.py:87]: can you reduce number of return statements to one?

**Review by marina-kaz — COMMENTED** [review]
Good start!

File: lab_4_fill_words_by_ngrams/main.py (hunk 3 above)
Comment (Dzianna7) [reply lab_4_fill_words_by_ngrams/main.py:83]: fixed

File: lab_4_fill_words_by_ngrams/main.py (hunk 4 above)
Comment (Dzianna7) [reply lab_4_fill_words_by_ngrams/main.py:88]: fixed

**Comment by vitazaytseva** [comment]
Well done! Good job!

**Pull request closed without merging**